Source/Scene/ModelExperimental/MaterialPipelineStage.js

```diff
@@ -0,0 +1,485 @@
+  shaderBuilder.addDefine(
+    transformDefine,
+    undefined,
+    ShaderDestination.FRAGMENT
+  );
+
+  // Add a uniform for the transformation matrix
+  var transformUniformName = uniformName + "Transform";
+  shaderBuilder.addUniform(
+    "mat3",
+    transformUniformName,
+    ShaderDestination.FRAGMENT
+  );
+  uniformMap[transformUniformName] = function () {
+    return textureReader.transform;
+  };
+}
+
+/**
+ * Process a single texture and add it to the shader and uniform map.
+ *
+ * @param {ShaderBuilder} shaderBuilder The shader builder to modify
+ * @param {Object.<String, Function>} uniformMap The uniform map to modify.
+ * @param {ModelComponents.TextureReader} textureReader The texture to add to the shader
+ * @param {String} uniformName The name of the sampler uniform such as <code>u_baseColorTexture</code>
+ * @param {String} defineName The name of the texture for use in the defines, minus any prefix or suffix. For example, "BASE_COLOR" or "EMISSIVE"
+ *
+ * @private
+ */
+function processTexture(
+  shaderBuilder,
+  uniformMap,
+  textureReader,
+  uniformName,
+  defineName,
+  defaultTexture
+) {
+  // Add a uniform for the texture itself
+  shaderBuilder.addUniform(
+    "sampler2D",
+    uniformName,
+    ShaderDestination.FRAGMENT
+  );
+  uniformMap[uniformName] = function () {
+    return defaultValue(textureReader.texture, defaultTexture);
+  };
+
+  // Add a #define directive to enable using the texture in the shader
+  var textureDefine = "HAS_" + defineName + "_TEXTURE";
+  shaderBuilder.addDefine(textureDefine, undefined, ShaderDestination.FRAGMENT);
+
+  // Add a #define to tell the shader which texture coordinates varying to use.
+  var texCoordIndex = textureReader.texCoord;
+  var texCoordVarying = "v_texCoord_" + texCoordIndex;
+  var texCoordDefine = "TEXCOORD_" + defineName;
+  shaderBuilder.addDefine(
+    texCoordDefine,
+    texCoordVarying,
+    ShaderDestination.FRAGMENT
+  );
+
+  // Some textures have matrix transforms (e.g. for texture atlases). Add those
+  // to the shader if present.
+  var textureTransform = textureReader.transform;
+  if (
+    defined(textureTransform) &&
+    !Matrix3.equals(textureTransform, Matrix3.IDENTITY)
+  ) {
+    processTextureTransform(
+      shaderBuilder,
+      uniformMap,
+      textureReader,
+      uniformName,
+      defineName
+    );
+  }
+}
+
+function processMaterialUniforms(
+  material,
+  uniformMap,
+  shaderBuilder,
+  defaultTexture,
+  defaultNormalTexture,
+  defaultEmissiveTexture
+) {
+  var emissiveTexture = material.emissiveTexture;
+  if (defined(emissiveTexture)) {
+    processTexture(
+      shaderBuilder,
+      uniformMap,
+      emissiveTexture,
+      "u_emissiveTexture",
+      "EMISSIVE",
+      defaultEmissiveTexture
+    );
+  }
+
+  var emissiveFactor = material.emissiveFactor;
+  if (
+    defined(emissiveFactor) &&
+    !Cartesian3.equals(emissiveFactor, Cartesian3.ZERO)
+  ) {
```

Comment:
Ah, neat - I didn't see that `ModelComponents` defines all these. Since `ModelComponents` is supposed to be format agnostic, I'm curious if other 3D formats that do PBR also have these same defaults... or is it in the nature of PBR that all these defaults are always the same?

Reply:
I would imagine they would be the same or similar, PBR metallic roughness/specular glossiness workflows are pretty standard.

See [This guide](https://substance3d.adobe.com/tutorials/courses/the-pbr-guide-part-1) if you want to learn more about PBR